### DISCUSSION TITLE
How can I install tools with completions?

### DISCUSSION THREAD

**Discussion by kamontat** [jdx/mise#3757] [Q&A]
For example, [fd](https://github.com/sharkdp/fd) command contains [_fd](https://github.com/sharkdp/fd/tree/master/contrib/completion/_fd) file to use as autocompletions on zsh. How can I use it?

**Comment by jdx** [comment]
we don't currently have anything for this. I don't think it's possible since none of the backends mise relies on themselves contain completion information.

I think if you'd like to see this the best way would be to add the functionality to aqua first.

**Reply by hkalbasi** [reply]
@jdx What about having a field like `completion_command = "tool --completions {bash|fish|nu}` in the registry? All backends are capable of installing the binary, and for most tools the binary can generate its completions for some shells.

**Comment by kamontat** [comment]
@jdx Can we use hook to copy completion file from tool directory to specific completion directory?

**Reply by syhol** [reply]
There's about 9 backends for mise. Each backend downloads tools differently but almost none check out the tools repo, so there will be no local _fd file to copy.

I agree, completions for mise managed tools is a problem, but it's also a problem for all tools that aren't managed by a package manager like brew or apt (which do distribute completions)

Compare this to any other dev tools manager and you'll have the same problem. So I think this should be solved outside of mise. I would love to work on a project to handle this, but I don't have the time right now.

Until a better solution is in place, consider using the fish shell, which comes with many completions pre-installed.

**Reply by reitzig** [reply]
Many commands can generate their completions, and that [seems to be the case for fd as well](https://github.com/sharkdp/fd/issues/1278#issuecomment-1459595133). So we an use e.g. an [postinstall hook](https://mise.jdx.dev/hooks.html#preinstall-postinstall-hook) to install the completion.

**Comment by lboshell** [comment]
This is an interesting question. I personally use [chezmoi](https://www.chezmoi.io/) for this, which makes sense to me since I see this falling inside the "my personal configuration/dotfiles" problem, and not so much as part of mise's domain. The way to do it is usually one of two scenarios:

- If the tool includes a completion script in their repository, I add it to my dotfiles setup using [.chezmoiexternal.toml](https://www.chezmoi.io/user-guide/include-files-from-elsewhere/). For example, this is a fragment of my `~/.local/share/chezmoi/.chezmoiexternal.toml`:

```toml
[".config/fish/completions/buku.fish"]
type = "file"
url = "https://raw.githubusercontent.com/jarun/buku/refs/heads/master/auto-completion/fish/buku.fish"

[".config/fish/completions/eza.fish"]
type = "file"
url = "https://raw.githubusercontent.com/eza-community/eza/refs/heads/main/completions/fish/eza.fish"
```

This adds the completion scripts of two tools (buku and eza) for my shell (fish) directly from their repos.

- Maybe the tool's repo doesn't include completion scripts, but if the tool has a way to generate them from their CLI, I use [chezmoi scripts](https://www.chezmoi.io/user-guide/use-scripts-to-perform-actions/) to create them in the right place. These scripts can be setup to run only once every time the script itself changes, so it's an effective and efficient way to build my dotfiles structure (including completion files) and keep it up-to-date. For example, this is a script I have as `~/.local/share/chezmoi/private_dot_config/private_fish/completions/run_onchange_mise.fish`:

```fish
#! /usr/bin/env fish

echo "Creating completions for mise"
mise completion fish > ~/.config/fish/completions/mise.fish
```

I haven't come across the need to dynamically load/unload completion scripts when I enter a workspace/directory, but I think it can probably be done with the current functionality offered by `mise`+`chezmoi`.

**Reply by andrewaylett** [reply]
One of the nice things about `fish` is that it doesn't need the completions file to be static — it'll only run the content once, and only when it first needs a completion, so it's usually OK to have the tool dynamically generate its completions.

Which means that rather than an on-change script, I can drop in a file like [this one for JJ](https://github.com/andrewaylett/dotfiles/blob/main/dot_config/fish/completions/jj.fish) with content:

```fish
COMPLETE=fish jj | source
```

Fish itself keeps an array called `$fish_complete_path` which is a list of paths it'll look in for completions.  If I add my project completions directory to the head of that array, Fish will load the completion on-demand.  What I've not managed to work out is a way to mutate that array via mise -- obviously mise will set PATH, but I'm not seeing any way to also configure other PATH-like variables?

**Comment by reitzig** [comment]
Here's what I've come up with for fish:

- In the project `mise.toml`, set an env var to the target subdirectory for completions:
  ```toml
  LOCAL_FISH_COMPLETION_DIR = "{{config_root}}/.config/fish/completions"
  ```
- For each tool, use a `postinstall` hook to write its completions to a file in that subdirectory:
  ```toml
  postinstall = 'lefthook completion fish > "${LOCAL_FISH_COMPLETION_DIR}/lefthook.fish"'
  ```
- As user-level fish config, declare an [`--on-variable` function](https://fishshell.com/docs/current/cmds/function.html#:~:text=%2Dv%20VARIABLE_NAME%20or%20%2D%2Don%2Dvariable%20VARIABLE_NAME) that sources the completions from that subdirectory.

That last bit is key: I wasn't able to find a way to change the current shell's context/environment from a mise task. Everything I do in a task dies with its subshell/subprocess. By having the shell indirectly pick up on the event "entered a directory with local completions" via an env var being changed (set, really), I was able to work around that just enough to solve _this_ use case.

MWE: [reitzig/2885464e](https://gist.github.com/reitzig/2885464e5c2fcc0f84bda9e1dae4f7b1)

In my user-level `~/.config/mise/config.toml`, I do a similar thing but with

```toml
FISH_COMPLETION_DIR = "{{env.HOME}}/.config/fish/completions"
```

Since those completions reside in the default completion directory, fish picks them up by default and we don't have to source them; hence the different env var name.

_What this does:_ It works! When I enter such a project directory, the completions are available in the current shell session. By design, two different shell sessions in two different projects would be able to use different completions for different versions of the same tool.

_What this doesn't do:_ 

- Load completions in`postinstall`. Don't see how. Workaround is to re-enter the project directory.
- Unload completions when leaving the project directory. I don't see how. _But_ it probably doesn't matter: completions for programs not on the path don't hurt, do they?
- Clean up -- completion files never get removed. (potential performance impact, but otherwise probably fine)

_What I'd like to have_:

- A `postselfupdate` hook to update completions of `mise` itself.
- _Some_ way to nudge the shell from mise tasks or hooks, e.g. by changing an env var. (With this, I could do away with the "doesn't do"s above).
- An `uninstall` hook for _removing_ completions files.

_Credits:_ [pfgray/fish-completion-sync](https://github.com/pfgray/fish-completion-sync) -- I chose `source` over injecting additional files into fish's mechanism because it seemed simpler, but I got the idea for using an `--on-variable` function there.

**Reply by reitzig** [reply]
My approach could be simplified by running tasks directly in hooks: #3314

**Comment by alikefia** [comment]
Hello there,
I came across the same need and got it work via hooks (experimental)

```
[tools]
typst = "latest"
tinymist = "latest"

[[hooks.enter]]
shell = "bash"
script = "source <(typst completions bash)"
```

**Reply by davidpoblador** [reply]
@kamontat I agree the proper fix would be to have aqua support this. In the meantime I have stitched something together that does the trick: https://github.com/alltuner/mise-completions-sync

**Reply by smoebody** [reply]
this also works with zsh:
```toml
[tools]
hcloud = "latest"

[hooks.enter]
shell = "zsh"
script = "source <(hcloud completion zsh)"
```

**Comment by TyceHerrman** [comment]
Issue for shell completion support in aqua for folks that want to follow/upvote: https://github.com/aquaproj/aqua/issues/722

**Comment by andrewaylett** [comment]
I've got something working with fish:

```
[[hooks.enter]]
shell = "fish"
script = """
set -gp fish_complete_path $MISE_PROJECT_ROOT/.mise/completions/fish
source $MISE_PROJECT_ROOT/.mise/completions/fish/*.fish
"""

[[hooks.leave]]
shell = "fish"
script = """
set -l idx (contains -i -- $MISE_PROJECT_ROOT/.mise/completions/fish $fish_complete_path)
and set -e fish_complete_path[$idx]
for f in $MISE_PROJECT_ROOT/.mise/completions/fish/*.fish
    complete -e -c (basename -s .fish $f)
end
"""
```

Drop your completions in the relevant directory, and they'll be both added and removed automagically.  And if you add a completion within the session, it'll work too.
